pkg/processor/k8sprocessor/kube/owner.go

```diff
@@ -505,12 +504,43 @@ func (op *OwnerCache) genericEndpointSliceOp(obj interface{}, serviceFunc func(p
 		return
 	}
 
-	epLabels := endpointSlice.GetLabels()
-	serviceName := epLabels[endpointSliceServiceLabel] // see: https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#ownership
+	serviceName := getServiceName(endpointSlice)
 
 	for _, endpoint := range endpointSlice.Endpoints {
 		if endpoint.TargetRef != nil && endpoint.TargetRef.Kind == "Pod" {
-			serviceFunc(endpoint.TargetRef.Name, serviceName)
+			podName := endpoint.TargetRef.Name
+			serviceFunc(podName, serviceName)
+		}
+	}
+}
+
+func (op *OwnerCache) updateEndpointSlice(oldObj interface{}, newObj interface{}) {
+	// for updates, we're guaranteed the objects will be the right type
+	oldEndpointSlice := oldObj.(*discovery_v1.EndpointSlice)
+	newEndpointSlice := newObj.(*discovery_v1.EndpointSlice)
+
+	// add the new endpointslice first, the logic is the same
+	op.cacheEndpointSlice("EndpointSlice", newObj)
+
+	// we also need to remove the Service from Pods which were deleted from the endpointslice
+	serviceName := getServiceName(newEndpointSlice)
+
+	newPodNames := []string{}
+	for _, endpoint := range newEndpointSlice.Endpoints {
+		if endpoint.TargetRef != nil && endpoint.TargetRef.Kind == "Pod" {
+			podName := endpoint.TargetRef.Name
+			newPodNames = append(newPodNames, podName)
+		}
+	}
+
+	// for each Pod name which was in the old slice, but not in the new slice, schedule a delete
+	for _, endpoint := range oldEndpointSlice.Endpoints {
+		if endpoint.TargetRef != nil && endpoint.TargetRef.Kind == "Pod" {
+			podName := endpoint.TargetRef.Name
+			if slices.Index(newPodNames, podName) == -1 {
+				// not a deferred delete, as this is a dynamic property which can change often
+				op.deleteServiceFromPod(podName, serviceName)
```

Comment:
Is a call to `deleteServiceFromPod` going to ensure that our cache doesn't have any stale entries? So my understanding here is that whenever any EndpointSlice for a service is updated this method gets called with the params as mentioned here - https://pkg.go.dev/k8s.io/client-go@v0.29.0/tools/cache#ResourceEventHandlerFuncs.OnUpdate

Reply:
Yes, that's the idea. An update to an EndpointSlice can mean that some Pods were removed from the Service, and this is what we do here.